### PR TITLE
update homog. engine - mpi parallelization

### PR DESCRIPTION
**Pull request by vlukes**

### CODE REVIEW

**Comment by rc** [comment]
All Travis CI tests now fail, because mpi4py is not in the compulsory requirements, and `.travis.yml` does not tell to install it: 
```
<<< directory: tests, test files: 99
<<< [1] tests/test_assembling.py
[1] sfepy: left over: ['nm', '__builtins__', '_filename', '__file__', 'sps', 'absolute_import', '__package__', '__doc__', '__name__', 'TestCommon', 'verbose']
>>> test instance prepared (4 test(s))
>>> <type 'exceptions.NameError'>
Traceback (most recent call last):
  File "run_tests.py", line 234, in <module>
    sys.exit(main())
  File "run_tests.py", line 226, in main
    run_tests(stats, dirpath, filenames)
  File "run_tests.py", line 144, in run_tests
    stats[0] + 1)
  File "run_tests.py", line 108, in run_test
    ifile=ifile)
  File "/home/travis/build/sfepy/sfepy/sfepy/base/testing.py", line 67, in run
    ret = test_method()
  File "/home/travis/build/sfepy/sfepy/tests/test_assembling.py", line 64, in test_assemble_matrix
    from sfepy.discrete.common.extmods.assemble import assemble_matrix
  File "/home/travis/build/sfepy/sfepy/sfepy/discrete/__init__.py", line 3, in <module>
    from sfepy.discrete.common.domain import Domain
  File "/home/travis/build/sfepy/sfepy/sfepy/discrete/common/__init__.py", line 2, in <module>
    from .fields import Field
  File "/home/travis/build/sfepy/sfepy/sfepy/discrete/common/fields.py", line 7, in <module>
    import sfepy.base.multiproc as multi
  File "/home/travis/build/sfepy/sfepy/sfepy/base/multiproc.py", line 4, in <module>
    import sfepy.base.multiproc_mpi as multiproc_mpi
  File "/home/travis/build/sfepy/sfepy/sfepy/base/multiproc_mpi.py", line 22, in <module>
    class MPILogFile(MPI.File):
NameError: name 'MPI' is not defined
```

Could you rearrange the imports, so that the code runs without mpi4py?

File: sfepy/base/multiproc.py [hunk 1]
```diff
@@ -1,116 +1,41 @@
...
+                      multiproc_threads.use_multiprocessing
 
-global_multiproc_dict = {}
+multiprocessing_mode = None
+multiprocessing_module = None
 
+def get_multiproc(mpi=False):
+    global multiprocessing_mode, multiprocessing_module
+    try_threads = True
+    miltiproc, mode = None, None
```
Comment (rc) [inline sfepy/base/multiproc.py]: miltiproc -> multiproc

File: sfepy/base/multiproc_threads.py [hunk 2]
```diff
@@ -0,0 +1,121 @@
...
+            return None
+
+    def put(self, value):
+        self.queue.put(value)
+
+
+def get_manager():
+    """
+    Get the multiprocessing manager. If not in the global cache,
+    create the one.
```
Comment (rc) [inline sfepy/base/multiproc_threads.py]: create a new instance.

File: sfepy/base/multiproc_threads.py [hunk 3]
```diff
@@ -0,0 +1,121 @@
...
+    """
+    if use_multiprocessing and 'manager' not in global_multiproc_dict:
+        global_multiproc_dict['manager'] = Manager()
+
+    return global_multiproc_dict['manager']
+
+
+def get_mpdict_value(mode, key, clear=False):
+    """
+    Get the item form the global multiprocessing cache.
```
Comment (rc) [inline sfepy/base/multiproc_threads.py]: form -> from

File: sfepy/discrete/common/fields.py [hunk 4]
```diff
@@ -4,9 +4,10 @@
 import numpy as nm
 
 from sfepy.base.base import output, iter_dict_of_lists, Struct, basestr
-from sfepy.base.multiproc import use_multiprocessing, managers
+import sfepy.base.multiproc as multi
```
Comment (rc) [inline sfepy/discrete/common/fields.py]: Try moving this import into the functions that need it to break the hard dependence on mpi.

File: simple_homog_mpi.py [hunk 5]
```diff
@@ -0,0 +1,133 @@
+#!/usr/bin/env python
+# 06.06.2017, c
+"""
+Solve a coupled two-scale problem in parallel. One comptational node is solving
```
Comment (rc) [inline simple_homog_mpi.py]: comptational -> computational

File: simple_homog_mpi.py [hunk 6]
```diff
@@ -0,0 +1,133 @@
...
+a macroscopic eqution while the others are solvig local microscopic problems
```
Comment (rc) [inline simple_homog_mpi.py]: (eqution -> equation) while the other nodes are (solvig -> solving) local microscopic problems

File: simple_homog_mpi.py [hunk 7]
```diff
@@ -0,0 +1,133 @@
...
+and homogenized coefficients.
+
+Run this script as:
+
+    mpiexec -n 4 simple_homog_mpi.py examples/homogeization/nonlinear_hyperelastic_mM.py
```
Comment (rc) [inline simple_homog_mpi.py]: typo in the path

File: simple_homog_mpi.py [hunk 8]
```diff
@@ -0,0 +1,133 @@
+#!/usr/bin/env python
+# 06.06.2017, c
+"""
+Solve a coupled two-scale problem in parallel. One comptational node is solving
+a macroscopic eqution while the others are solvig local microscopic problems
+and homogenized coefficients.
+
+Run this script as:
```
Comment (rc) [inline simple_homog_mpi.py]: Run this script as::

**Comment by rc** [comment]
The PR introduces a new top-level script, so add it into sfepy-run, please.

File: simple_homog_mpi.py (hunk 8 above)
Comment (vlukes) [inline simple_homog_mpi.py]: fixed

File: simple_homog_mpi.py (hunk 7 above)
Comment (vlukes) [inline simple_homog_mpi.py]: fixed

File: simple_homog_mpi.py (hunk 6 above)
Comment (vlukes) [inline simple_homog_mpi.py]: fixed

File: simple_homog_mpi.py (hunk 5 above)
Comment (vlukes) [inline simple_homog_mpi.py]: fixed

File: sfepy/base/multiproc_threads.py (hunk 3 above)
Comment (vlukes) [inline sfepy/base/multiproc_threads.py]: fixed

File: sfepy/discrete/common/fields.py (hunk 4 above)
Comment (vlukes) [inline sfepy/discrete/common/fields.py]: fixed

**Comment by vlukes** [comment]
Imports are now rearranged to run it without mpi4py.

**Comment by rc** [comment]
Thanks, the tests pass now, so I will merge it.

**Comment by rc** [comment]
I have tried running `mpiexec -n 4 simple_homog_mpi.py examplehomogenization/nonlinear_hyperelastic_mM.py` and got:
```
...
  File ".../sfepy-git/examples/homogenization/nonlinear_hyperelastic_mM.py", line 79, in get_homog_mat
    iteration=problem.iiter, **kwargs)
  File ".../sfepy-git/sfepy/homogenization/micmac.py", line 97, in get_homog_coefs_nonlinear
    if multi_mpi is not None:
UnboundLocalError: local variable 'multi_mpi' referenced before assignment
```

Is it a problem of my installation, or a bug?

**Comment by vlukes** [comment]
You must add `'use_mpi': True` to the options in `nonlinear_homogenization.py`

**Comment by rc** [comment]
OK, that works. Thanks! Let us merge the PR in this state, polishing may come later.

**Pull request merged**